src/test/java/ch/heigvd/pro/b04/questions/ServerQuestionTest.java

```diff
@@ -139,10 +141,75 @@ public void testParticipantCannotInsertOrUpdateOrDeleteQuestion() {
 
     lenient().when(modoRepo.findByToken("t1")).thenReturn(Optional.empty());
     lenient().when(participantRepository.findByToken("t1")).thenReturn(Optional.of(aloy));
-    //lenient().when(repo.delete(Mockito.any())).thenReturn(new ServerMessage("success"));
 
     assertThrows(WrongCredentialsException.class, () -> cc.insert(1, 123, "t1", c2));
     assertThrows(WrongCredentialsException.class, () -> cc.updateQuestion(1, 123, 1, "t1", c2));
     assertThrows(WrongCredentialsException.class, () -> cc.deleteQuestion(1, 123, 1, "t1"));
   }
+
+  @Test
+  public void testModeratorCanInsertOrUpdateQuestionOnlyWithRightAccess()
+      throws ResourceNotFoundException, WrongCredentialsException {
+    ClientQuestion c1 = ClientQuestion.builder()
+        .title("Do you dream of Scorchers ?").build();
+
+    ServerPoll pollTemp = ServerPoll.builder()
+        .idPoll(ServerPollIdentifier.builder().idPoll(123).build())
+        .build();
+
+    Moderator aloy = Moderator.builder()
+        .idModerator(1)
+        .username("aloy")
+        .secret("chieftain")
+        .pollSet(Set.of(pollTemp))
+        .build();
+    pollTemp.getIdPoll().setIdxModerator(aloy);
+
+    Moderator ikrie = Moderator.builder()
+        .idModerator(2)
+        .username("ikrie")
+        .secret("banuk")
+        .build();
+
+    when(modoRepo.findByToken("t1")).thenReturn(Optional.of(aloy));
+    when(modoRepo.findByToken("t2")).thenReturn(Optional.of(ikrie));
+    when(pollRepo.findById(Mockito.any())).thenReturn(Optional.empty());
+    when(pollRepo.findById(
+        ServerPollIdentifier.builder()
+        .idPoll(123).idxModerator(aloy).build()
+    )).thenReturn(Optional.of(pollTemp));
+
+    //Moderator who owns the poll
+    assertDoesNotThrow(() -> cc.insert(1, 123, "t1", c1));
+    //Moderator who does not own the poll
+    assertThrows(ResourceNotFoundException.class, () ->cc.insert(2, 123, "t2", c1));
+
+    ServerQuestion q2 = ServerQuestion.builder()
+        .idServerQuestion(ServerQuestionIdentifier.builder()
+            .idServerQuestion(2)
+            .build())
+        .title("Do you love the Frostclaws ?")
+        .build();
+
+    ServerPoll pollTemp2 = ServerPoll.builder()
+        .idPoll(ServerPollIdentifier.builder().idPoll(223).build())
+        .pollServerQuestions(Set.of(q2))
+        .build();
+
+    q2.getIdServerQuestion().setIdxPoll(pollTemp2);
```

Comment:
Please use the `Builder` (and reorder code if necessary).

Reply:
We can't reorganize code here, that's what I said on Telegram: using exclusively the Builders, the snake bites his tail. We have to set idxPoll after.